builtin/v12/migration/miner.go

```diff
@@ -405,3 +499,97 @@ func migrateSectorInfo(sectorInfo miner11.SectorOnChainInfo) *miner12.SectorOnCh
 		SimpleQAPower:         sectorInfo.SimpleQAPower,
 	}
 }
+
+func migrateDeadlineSectorsWithDiff(ctx context.Context, store adt11.Store, inRoot cid.Cid, prevInRoot cid.Cid, prevOutRoot cid.Cid) (cid.Cid, error) {
+	// we have previous work, but the AMT has changed -- diff them
+	diffs, err := amt.Diff(ctx, store, store, prevInRoot, inRoot, amt.UseTreeBitWidth(miner11.SectorsAmtBitwidth))
+	if err != nil {
+		return cid.Undef, xerrors.Errorf("failed to diff old and new Sector AMTs: %w", err)
+	}
+
+	inSectors, err := miner11.LoadSectors(store, inRoot)
+	if err != nil {
+		return cid.Undef, xerrors.Errorf("failed to load inSectors: %w", err)
+	}
+
+	prevOutSectors, err := miner12.LoadSectors(store, prevOutRoot)
+	if err != nil {
+		return cid.Undef, xerrors.Errorf("failed to load prevOutSectors: %w", err)
+	}
+
+	for _, change := range diffs {
+		switch change.Type {
+		case amt.Remove:
+			if err := prevOutSectors.Delete(change.Key); err != nil {
+				return cid.Undef, xerrors.Errorf("failed to delete sector from prevOutSectors: %w", err)
+			}
+		case amt.Add:
+			fallthrough
+		case amt.Modify:
+			sectorNo := abi.SectorNumber(change.Key)
+			info, found, err := inSectors.Get(sectorNo)
+			if err != nil {
+				return cid.Undef, xerrors.Errorf("failed to get sector %d in inSectors: %w", sectorNo, err)
+			}
+
+			if !found {
+				return cid.Undef, xerrors.Errorf("didn't find sector %d in inSectors", sectorNo)
+			}
+
+			if err := prevOutSectors.Set(change.Key, migrateSectorInfo(*info, store)); err != nil {
+				return cid.Undef, xerrors.Errorf("failed to set migrated sector %d in prevOutSectors", sectorNo)
+			}
+		}
+	}
+
+	return prevOutSectors.Root()
+}
+
+func migrateDeadlineSectorsFromScratch(ctx context.Context, store adt11.Store, inArray *adt11.Array) (*adt12.Array, error) {
+	outArray, err := adt12.MakeEmptyArray(store, miner12.SectorsAmtBitwidth)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to construct new sectors array: %w", err)
+	}
+
+	var sectorInfo miner11.SectorOnChainInfo
+	if err = inArray.ForEach(&sectorInfo, func(k int64) error {
+		return outArray.Set(uint64(k), migrateSectorInfo(sectorInfo, store))
+	}); err != nil {
+		return nil, err
+	}
+
+	return outArray, err
+}
+
+func addSectorNumberToDealIdHAMT(xap *builtin.ActorTree, sectorInfo miner11.SectorOnChainInfo, store adt11.Store) error {
+	//sectorInfo.DealIDs,sectorInfo.SectorNumber
+	// - todo confirm types beint stored is correct
+
+	cborDealIDs := market.SectorDealIDs{DealIDs: sectorInfo.DealIDs}
+	err := xap.Map.Put(abi.IntKey(int64(sectorInfo.SectorNumber)), &cborDealIDs)
```

Comment:
nit: `xap.Map.Put(abi.IntKey(int64(sectorInfo.SectorNumber)), &market.SectorDealIDs{DealIDs: sectorInfo.DealIDs})` since IIRC go lets you get away with construction and serialization in one line.

Reply:
#ffbf314 done